main/utils/trophy.c

```diff
@@ -0,0 +1,1262 @@
+
+        // Load sprite
+        if (tw->trophyData.image != NO_IMAGE_SET)
+        {
+            loadWsg(tw->trophyData.image, &tw->image, true);
+        }
+
+        // Check if final trophy
+        if (final)
+        {
+            // Save to NVS
+            trophySystem.plat.currentVal = 1;
+            _save(&trophySystem.plat, 1);
+            _setPoints(1000 - (trophySystem.numTrophiesScore * _genPoints(0)));
+            _saveLatestWin(&trophySystem.plat);
+            // Add to end of queue
+            push(&trophySystem.trophyQueue, &trophySystem.plat);
+        }
+
+        // Set system and this index to active
+        if (!trophySystem.active)
+        {
+            trophySystem.active    = true;
+            trophySystem.animTimer = 0;
+        }
+    }
+}
+
+void trophyUpdateMilestone(trophyData_t t, int newVal, int threshold)
+{
+    trophyDataWrapper_t* tw = heap_caps_calloc(1, sizeof(trophyDataWrapper_t), MALLOC_CAP_8BIT);
+    _load(tw, t);
+
+    // Check if completed
+    if (newVal >= t.maxVal)
+    {
+        trophyUpdate(t, newVal, true);
+        heap_caps_free(tw);
+        return;
+    }
+
+    // Check if past a milestone for the first time
+    // Verify Percentage isn't out of bounds
+    if (threshold < 1 || threshold >= 100)
+    {
+        threshold = DEFAULT_MILESTONE;
+    }
+    int32_t prev = (tw->currentVal * 100 / t.maxVal) / threshold; // Intentionally truncates to an integer
+    int32_t new  = (newVal * 100 / t.maxVal) / threshold;
+    trophyUpdate(t, newVal, prev < new);
+    heap_caps_free(tw);
+}
+
+int32_t trophyGetSavedValue(trophyData_t t)
+{
+    int32_t val;
+    char buffer[MAX_NVS_KEY_LEN];
+    _truncateStr(buffer, t.title, MAX_NVS_KEY_LEN);
+    if (readNamespaceNvs32(trophySystem.data->settings->namespaceKey, buffer, &val))
+    {
+        return val;
+    }
+    return 0;
+}
+
+void trophySetChecklistTask(trophyData_t t, int32_t flag, bool set, bool drawUpdate)
+{
+    // If not a checklist, abort
+    if (t.type != TROPHY_TYPE_CHECKLIST)
+    {
+        return;
+    }
+
+    // Load
+    trophyDataWrapper_t* tw = heap_caps_calloc(1, sizeof(trophyDataWrapper_t), MALLOC_CAP_8BIT);
+    _load(tw, t);
+
+    int32_t newVal = tw->currentVal;
+    setBitFlag(&newVal, flag, set);
+
+    // Run Update
+    trophyUpdate(t, newVal, drawUpdate);
+
+    heap_caps_free(tw);
+}
+
+void trophyClear(trophyData_t t)
+{
+    if (trophySystem.plat.currentVal == 1)
+    {
+        _save(&trophySystem.plat, 0);
+        trophySystem.plat.currentVal = 0;
+        int rmPoints                 = 1000 - (trophySystem.numTrophiesScore * _genPoints(0));
+        _setPoints(rmPoints * -1);
+    }
+
+    trophyDataWrapper_t tw = {};
+    _load(&tw, t);
+    // Reset points
+    if (tw.currentVal >= t.maxVal)
+    {
+        _setPoints(_genPoints(t.difficulty) * -1);
+    }
+
+    // Finalize
+    _save(&tw, 0);
+}
+
+// Helpers
+
+bool checkBitFlag(int32_t flags, int8_t idx)
```

Comment:
This could be declared `inline`. It's similar to a macro function (see below), but has all the safety and type checking of a function.
```c
#define CHECK_BIT_FLAG(flags, idx) ((flags & (1 << idx)) != 0)
```
https://gcc.gnu.org/onlinedocs/gcc/Inline.html